lib/fileset.nix

```diff
@@ -0,0 +1,993 @@
+    let
+      recurse = focusPath: tree:
+        if tree == "directory" && expand || isAttrs tree then
+          let
+            expanded = _directoryEntries focusPath tree;
+            transformedSubtrees = mapAttrs (name: recurse (append focusPath name)) expanded;
+            values = attrValues transformedSubtrees;
+          in
+          if all (value: value == "emptyDir") values then
+            "emptyDir"
+          else if all (value: isNull value || value == "emptyDir") values then
+            null
+          else if !expand && all (value: isString value || value == "emptyDir") values then
+            "directory"
+          else
+            mapAttrs (name: value: if value == "emptyDir" then null else value) transformedSubtrees
+        else
+          tree;
+      result = recurse base tree;
+    in
+    if result == "emptyDir" then
+      null
+    else
+      result;
+
+  _prettyTreeSuffix = tree:
+    if isAttrs tree then
+      ""
+    else if tree == "directory" then
+      " (recursive directory)"
+    else
+      " (${tree})";
+
+  # Pretty-print all files included in the file set.
+  # Type: (b -> String -> b) -> b -> Path -> FileSet -> b
+  _prettyFoldl' = f: start: base: tree:
+    let
+      traceTreeAttrs = start: indent: tree:
+        # Nix should really be evaluating foldl''s second argument before starting the iteration
+        # See the same problem in Haskell:
+        # - https://stackoverflow.com/a/14282642
+        # - https://gitlab.haskell.org/ghc/ghc/-/issues/12173
+        # - https://well-typed.com/blog/90/#a-postscript-which-foldl
+        # - https://old.reddit.com/r/haskell/comments/21wvk7/foldl_is_broken/
+        seq start
+          (foldl' (prev: name:
+            let
+              subtree = tree.${name};
+
+              intermediate =
+                f prev "${indent}- ${name}${_prettyTreeSuffix subtree}";
+            in
+            if subtree == null then
+              # Don't print anything at all if this subtree is empty
+              prev
+            else if isAttrs subtree then
+              # A directory with explicit entries
+              # Do print this node, but also recurse
+              traceTreeAttrs intermediate "${indent}  " subtree
+            else
+              # Either a file, or a recursively included directory
+              # Do print this node but no further recursion needed
+              intermediate
+          ) start (attrNames tree));
+
+      intermediate =
+        if tree == null then
+          f start "${toString base} (empty)"
+        else
+          f start "${toString base}${_prettyTreeSuffix tree}";
+    in
+    if isAttrs tree then
+      traceTreeAttrs intermediate "" tree
+    else
+      intermediate;
+
+  # Coerce and normalise the bases of multiple file set values passed to user-facing functions
+  # Type: String -> [ { context :: String, value :: Any } ] -> { commonBase :: Path, trees :: [ <tree> ] }
+  _normaliseBase = function: list:
```

Comment:
For my own future reference for once I get to implementing the combinators (in particular `intersect`):

This function here should only be used for `union`, not for `difference` or `intersect`. In particular:
- For `union`, the resulting base path should be the common prefix of all the inputs base path, e.g. `union /foo /foo/bar` should give `/foo` as the base, this is what this function does
- For `intersect`, the resulting base path should be the path with the most components of the input bases, e.g. `intersect /foo /foo/bar` should give `/foo/bar` as the base. If the paths don't overlap, an empty set is returned.
  - TODO: What base should that have? E.g. `intersect /foo /bar`, should that give an error, pick `/foo` or `/bar` arbitrarily, or use `/`? Maybe the notion of a special empty file set value should be introduced.
- For `difference, the resulting base is the same as the base from the first argument, e.g. `difference /foo /` has base `/foo`.